gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeTableOperations.java

```diff
@@ -0,0 +1,229 @@
+        throw new CommitFailedException(e, "Failed to commit");
+      }
+      if (commitStatus == CommitStatus.UNKNOWN) {
+        throw new CommitStateUnknownException(e);
+      }
+    } finally {
+      try {
+        if (commitStatus == CommitStatus.FAILURE) {
+          LOG.warn("Failed to commit updates to table {}", tableName());
+          io().deleteFile(newMetadataLocation);
+        }
+      } catch (RuntimeException e) {
+        LOG.error(
+            "Failed to cleanup metadata file at {} for table {}",
+            newMetadataLocation,
+            tableName(),
+            e);
+      }
+    }
+  }
+
+  @Override
+  public String tableName() {
+    return String.format(
+        "%s.%s.%s", tableName.getCatalog(), tableName.getDatabase(), tableName.getTable());
+  }
+
+  @Override
+  public FileIO io() {
+    return fileIO;
+  }
+
+  private void createTable(String newMetadataLocation, TableMetadata metadata) {
+    LOG.debug("Creating a new Iceberg table: {}", tableName());
+    client.createTable(tableName, makeNewTable(metadata, newMetadataLocation));
+  }
+
+  /** Update table properties with concurrent update detection using etag. */
+  private void updateTable(
+      String oldMetadataLocation, String newMetadataLocation, TableMetadata metadata) {
+    Table table = client.getTable(tableName);
+    String etag = table.getEtag();
+    if (etag.isEmpty()) {
+      throw new ValidationException(
+          "Etag of legacy table %s is empty, manually update the table by BigLake API or"
+              + " recreate and retry",
+          tableName());
+    }
+    HiveTableOptions options = table.getHiveOptions();
+
+    // If `metadataLocationFromMetastore` is different from metadata location of base, it means
+    // someone has updated metadata location in metastore, which is a conflict update.
+    String metadataLocationFromMetastore =
+        options.getParametersOrDefault(METADATA_LOCATION_PROP, "");
+    if (!metadataLocationFromMetastore.isEmpty()
+        && !metadataLocationFromMetastore.equals(oldMetadataLocation)) {
+      throw new CommitFailedException(
+          "Base metadata location '%s' is not same as the current table metadata location '%s' for"
+              + " %s.%s",
+          oldMetadataLocation,
+          metadataLocationFromMetastore,
+          tableName.getDatabase(),
+          tableName.getTable());
+    }
+
+    try {
+      client.updateTableParameters(
+          tableName, buildTableParameters(newMetadataLocation, metadata), etag);
+    } catch (AbortedException e) {
+      if (e.getMessage().toLowerCase().contains("etag mismatch")) {
+        throw new CommitFailedException(
+            "Updating table failed due to conflict updates (etag mismatch)");
+      }
+    }
+  }
+
+  private Table makeNewTable(TableMetadata metadata, String metadataFileLocation) {
+    Table.Builder builder = Table.newBuilder().setType(Table.Type.HIVE);
+    builder
+        .getHiveOptionsBuilder()
```

Comment:
Is this necessary?  These options don't apply to Iceberg tables.

Reply:
Another way to use BigLake Metastore for Iceberg tables is by installing an HMS proxy exposing a local 9083 Thrift port. This proxy reads from the Metastore API and returns HMS tables. The benefit of this approach is that it works for all data engines that support HMS (without having to write a catalog plugin). We populate these fields because the Hive Iceberg catalog populates them. It makes these fields stored in BigLake Metastore, so the HMS proxy path can return exactly the same table as HMS does.

We received requests to develop more Iceberg catalog plugins (e.g., for Trino) for BigLake Metastore. After we have these catalog plugins, maybe we don't need to keep this HMS path any more. If it is OK, I'd like to keep the code, and we can remove them in future.